### PR TITLE
Revamp test selection for the example tests

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

After the revamp of the test_fetcher, the test collection of the example tests wasn't working anymore: if only an example is changed, the corresponding tests are not run. This PR adapts the example tests collection to the new test fetcher and makes sure they are appropriately run. It's more fine-grained than the previous approach which ran all example tests as soon as when a diff was discovered: here the tests are run when the modifications impact the test examples.

To see this in action:
- at the [first commit](https://app.circleci.com/pipelines/github/huggingface/transformers/65204) the diff only impacts repo utils, so only the test repo utils job is ran (no example or other test jobs).
- at the [second commit](https://app.circleci.com/pipelines/github/huggingface/transformers/65207) the diff has a change in the PyTorch `run_glue`, so the PyTorch example test job is ran, but only on the trainer examples (not the no_trainer ones since they are not touched).
- at the [third commit](https://app.circleci.com/pipelines/github/huggingface/transformers/65208), we remove the fake example modif and add a fake change in the Trainer, which impacts all examples.

cc @ydshieh for when you're back from vacation.

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
Looks good! Good reminder to have an XLA example running at some point.

File: .circleci/config.yml [hunk 1]
```diff
@@ -62,19 +68,6 @@ jobs:
...
-                      mv test_list.txt test_preparation/examples_test_list.txt
-                  else
-                      touch test_preparation/examples_test_list.txt
-                  fi
-            - run: |
-                  if [ -f filtered_test_list_cross_tests.txt ]; then
-                      mv filtered_test_list_cross_tests.txt test_preparation/filtered_test_list_cross_tests.txt
-                  else
-                      touch test_preparation/filtered_test_list_cross_tests.txt
-                  fi
```
Comment (ydshieh) [inline .circleci/config.yml:77]: Is any reason to remove this part regarding `filtered_test_list_cross_tests.txt`?

File: utils/tests_fetcher.py [hunk 2]
```diff
@@ -512,15 +513,40 @@ def print_tree_deps_of(module, all_edges=None):
...
+def init_test_examples_dependencies():
+    """
+    The test examples do not import from the examples (which are just scripts, not modules) so we need som extra
+    care initializing the dependency map there.
+    """
+    test_example_deps = {}
+    all_examples = []
+    for framework in ["flax", "pytorch", "tensorflow"]:
+        test_files = list((PATH_TO_EXAMPLES / framework).glob("test_*.py"))
+        all_examples.extend(test_files)
```
Comment (ydshieh) [inline utils/tests_fetcher.py:525]: Is the reason to include example test files (and return `all_examples`) only for testing purpose in `test_tests_fetcher.py`, and not really necessary for the line (below) `all_modules = list(PATH_TO_TRANFORMERS.glob("**/*.py")) + list(PATH_TO_TESTS.glob("**/*.py")) + examples`? (The `examples` part). See my comment there.

File: utils/tests_fetcher.py [hunk 3]
```diff
@@ -512,15 +513,40 @@ def print_tree_deps_of(module, all_edges=None):
...
     Create the dependency map from module/test filename to the list of modules/tests that depend on it (even
     recursively).
     """
     cache = {}
-    all_modules = list(PATH_TO_TRANFORMERS.glob("**/*.py")) + list(PATH_TO_TESTS.glob("**/*.py"))
+    example_deps, examples = init_test_examples_dependencies()
+    all_modules = list(PATH_TO_TRANFORMERS.glob("**/*.py")) + list(PATH_TO_TESTS.glob("**/*.py")) + examples
     all_modules = [str(mod.relative_to(PATH_TO_REPO)) for mod in all_modules]
     direct_deps = {m: get_module_dependencies(m, cache=cache) for m in all_modules}
+    direct_deps.update(example_deps)
```
Comment (ydshieh) [inline utils/tests_fetcher.py:549]: It seems we don't really need to include test example files  in `examples` (i.e. the above `all_examples`) for this `all_modules`. I guess it doesn't matter, and we can safely update `direct_deps` with `example_deps`. But logically, it seems a bit unusual to update (replace) the dictionary values here (which are lists) for the keys of `direct_deps` that are already in `example_deps`.

I don't mean we need to change anything, but a few comment(s) would be nice, even just in the line `all_examples.extend(test_files)`.

**Review by ydshieh — COMMENTED** [review]
Thank you for saving the examples!

2 or 3 nit questions from my side, but nothing big.

File: .circleci/config.yml (hunk 1 above)
Comment (sgugger) [reply .circleci/config.yml:77]: Probably an oversight, good catch!

File: utils/tests_fetcher.py (hunk 3 above)
Comment (sgugger) [reply utils/tests_fetcher.py:549]: I don't understand your comments. The keys in `example_deps` are not in `direct_deps`, as they are the text example files.

File: utils/tests_fetcher.py (hunk 2 above)
Comment (sgugger) [reply utils/tests_fetcher.py:525]: No it's just there is a custom logic with some files in the example folder being test files with weird imports and other being example files.

**Pull request merged**